### PR TITLE
8334147: Shenandoah: Avoid taking lock for disabled free set logging

### PR DESCRIPTION
**Pull request by earthling-amzn**
Not clean. Backport to only take lock for logging when log level is enabled.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issues
 * [JDK-8334147](https://bugs.openjdk.org/browse/JDK-8334147): Shenandoah: Avoid taking lock for disabled free set logging (**Enhancement** - P4)
 * [JDK-8341554](https://bugs.openjdk.org/browse/JDK-8341554): Shenandoah: Missing heap lock when updating usage for soft ref policy (**Bug** - P4)

### Reviewers
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Committer) ⚠️ Review applies to [275e76d9](https://git.openjdk.org/shenandoah-jdk21u/pull/113/files/275e76d98c05f9f4251ea4e195f8c0afdbf1eccc)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [275e76d9](https://git.openjdk.org/shenandoah-jdk21u/pull/113/files/275e76d98c05f9f4251ea4e195f8c0afdbf1eccc)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah-jdk21u.git pull/113/head:pull/113` \
`$ git checkout pull/113`

Update a local copy of the PR: \
`$ git checkout pull/113` \
`$ git pull https://git.openjdk.org/shenandoah-jdk21u.git pull/113/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 113`

View PR using the GUI difftool: \
`$ git pr show -t 113`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah-jdk21u/pull/113.diff">https://git.openjdk.org/shenandoah-jdk21u/pull/113.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah-jdk21u/pull/113#issuecomment-2387202678)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahGenerationalControlThread.cpp [hunk 1]
```diff
@@ -265,18 +263,15 @@ void ShenandoahGenerationalControlThread::run_service() {
...
+      heap->free_set()->log_status_under_lock();
 
-        // Notify Universe about new heap usage. This has implications for
-        // global soft refs policy, and we better report it every time heap
-        // usage goes down.
-        heap->update_capacity_and_used_at_gc();
+      // Notify Universe about new heap usage. This has implications for
+      // global soft refs policy, and we better report it every time heap
+      // usage goes down.
+      heap->update_capacity_and_used_at_gc();
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahGenerationalControlThread.cpp:271]: same question as above: does this need to be under lock?

**Review by ysramakrishna — APPROVED** [review]
Backport LGTM.

To Kelvin's question: I think we don't need the lock as these fields are updated only by the control thread (for consumption by jmx metrics and such). (I could be wrong and didn't check super-carefully.)

File: src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp [hunk 2]
```diff
@@ -179,18 +176,15 @@ void ShenandoahControlThread::run_service() {
...
+      heap->free_set()->log_status_under_lock();
 
-        // Notify Universe about new heap usage. This has implications for
-        // global soft refs policy, and we better report it every time heap
-        // usage goes down.
-        heap->update_capacity_and_used_at_gc();
+      // Notify Universe about new heap usage. This has implications for
+      // global soft refs policy, and we better report it every time heap
+      // usage goes down.
+      heap->update_capacity_and_used_at_gc();
```
[the remark this answers is not in the document]
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:184]: Yes, I looked at this. These two fields are defined in `CollectedHeap` and reflect `committed` and `used` for Shenandoah. They are only updated by the control thread and they are only read without the heap lock by the control thread or the vm thread to adjust the soft reference policy. The risk would be that if `used` somehow exceeded `committed`, but we have many assertions through `ShenandoahGeneration` to assert that this doesn't happen.

**Review by kdnilsen — APPROVED** [review]

**Pull request closed without merging**